File: Xamarin.Forms/xamarinMeetup2/MathModel.cs
// original performance comparition by Егор @Nagg
// https://habrahabr.ru/post/215329/
// source from https://github.com/EgorBo/Xamarin.Android-vs-Java/blob/master/ContactListXamarin/ContactListXamarin/Test1.cs

using System.Collections.Generic;
using System.Text;

namespace xamarinMeetup2
{
    public class MathModel
    {
        public string ArithmeticPerformance { get; set; }

        public string CollectionPerformance { get; set; }

        public string StringPerformance { get; set; }

        public void RunTests()
        {
            long arithmeticAverage = 0;
            long collectionsAverage = 0;
            long stringsAverage = 0;

            for (int i = 0; i < 5; i++)
            {
                arithmeticAverage += CalcArithmetic();
                collectionsAverage += CalcCollections();
                stringsAverage += CalcStrings();
            }

            ArithmeticPerformance = arithmeticAverage / 5 + "ms";
            CollectionPerformance = collectionsAverage / 5 + "ms";
            StringPerformance = stringsAverage / 5 + "ms";
        }

        long CalcArithmetic()
        {
            var timer = System.Diagnostics.Stopwatch.StartNew();

            int pans = 0;
            int primes = 0;
            double foo = 0;
            for (int i = 0; i < 12345; i++)
            {
                if (IsPandigital(i))
                    pans++;

                if (IsPrime(i))
                    primes++;

                foo += CalculateFoo(i);
            }

            timer.Stop();
            return timer.ElapsedMilliseconds;
        }

        long CalcCollections()
        {
            var timer = System.Diagnostics.Stopwatch.StartNew();

            var list = new List<FooContainer>(); //let's don't specify initial capacity
            for (int i = 0; i < 1000000; i++)
            {
                list.Add(new FooContainer(i, i));
            }
            var filteredList = new List<FooContainer>();
            for (int i = 0; i < list.Count; i++)
            {
                if (list[i].Sum % 2 == 0)
                {
                    filteredList.Add(list[i]);
                }
            }

            timer.Stop();
            return timer.ElapsedMilliseconds;
        }

        long CalcStrings()
        {
            var timer = System.Diagnostics.Stopwatch.StartNew();
            string result = "";
            var builder = new StringBuilder();
            for (int i = 0; i < 100000; i++)
            {
                builder.Append(i);
            }

            result = builder.ToString();
            result = result.Substring(1000);
            result = result.Replace("10", "");
            bool containsSubstring = result.Contains("123");
            string[] parts = result.Split('2');
            int length = parts.Length;
            timer.Stop();
            return timer.ElapsedMilliseconds;
        }

        bool IsPrime(int n)
        {
            if (n % 2 == 0) return false;
            for (int i = 3; i * i <= n; i += 2)
            {
                if (n % i == 0)
                    return false;
            }
            return true;
        }

        bool IsPandigital(int n)
        {
            int count = 0;
            int digits = 0;
            int digit;
            int bit;
            do
            {
                digit = n % 10;
                if (digit == 0)
                {
                    return false;
                }
                bit = 1 << digit;

                if (digits == (digits |= bit))
                {
                    return false;
                }

                count++;
                n /= 10;
            } while (n > 0);
            return (1 << count) - 1 == digits >> 1;
        }

        double CalculateFoo(int n)
        {
            double a = 1.0;
            for (int i = 0; i < n; i++)
            {
                a = a * 1.123456789101112 / n;
            }
            return a;
        }
    }

    struct FooContainer
    {
        private readonly int _x;
        private readonly int _y;

        public FooContainer(int x, int y)
        {
            _x = x;
            _y = y;
        }

        public int Sum
        {
            get { return _x + _y; }
        }
    }
}

File: Native.iOS/xm2 Native/MathModel.swift
//
//  MathModel.swift
//  xm2 Native
//
//  Created by Yury Nevalenny on 14/11/2016.
//  Copyright © 2016 Rahmobile. All rights reserved.
//

import Foundation

class MathModel {
    public var arithmeticPerformance: String = ""
    public var collectionsPerformance: String = ""
    public var stringsPerformance: String = ""
    let repetitions = 1
    
    public func calculate()
    {
        calcArithmetic ()
        calcCollections ()
        calcStrings ()
    }
    
    func calcArithmetic ()
    {
        var arithmetic: Double = 0
        
        for _ in 0..<repetitions
        {
            let start = DispatchTime.now()
            
            var pans : Int = 0
            var primes : Int = 0
            var foo : Double = 0
            
            for i in 0..<12345
            {
                if isPandigital(n: i) { pans+=1 }
                
                if isPrime(n: i) { primes+=1 }
                
                foo += calculateFoo(n: i)
            }
            
            let end = DispatchTime.now()
            let nanoTime = end.uptimeNanoseconds - start.uptimeNanoseconds
            let timeInterval = Double(nanoTime) / 1_000_000
            
            arithmetic += timeInterval
        }
        
        arithmetic /= Double(repetitions)
        arithmeticPerformance = String(Int(arithmetic)) + " ms"
    }
    
    func calcCollections ()
    {
        var collections: Double = 0
        
        for _ in 0..<repetitions
        {
            let start = DispatchTime.now()
            
            
            var list = Array<FooContainer>(); //let's don't specify initial capacity
            for i in 0..<1000000
            {
                var f = FooContainer()
                f._x=i
                f._y=i
                list.append(f)
            }
            
            var filteredList = Array<FooContainer>()
            for i in 0..<list.count
            {
                if list[i].Sum() % 2 == 0 {
                    filteredList.append(list[i])
                }
            }
            
            let end = DispatchTime.now()
            let nanoTime = end.uptimeNanoseconds - start.uptimeNanoseconds
            let timeInterval = Double(nanoTime) / 1_000_000
            
            collections += timeInterval
        }
        
        collections /= Double(repetitions)
        collectionsPerformance = String(Int(collections)) + " ms"
    }
    
    func calcStrings ()
    {
        var strings: Double = 0
        
        for _ in 0..<repetitions
        {
            let start = DispatchTime.now()
            
            var result = ""
            for i in 0..<100000
            {
                result.append(String(i))
            }
            
            let index = result.index(result.startIndex, offsetBy: 1000)
            result = result.substring(from: index)
            result = result.replacingOccurrences(of: "10", with: "")
            let containsSubstring = result.contains("123")
            let parts = result.components(separatedBy: "2")
            let length = parts.count
            
            let end = DispatchTime.now()
            let nanoTime = end.uptimeNanoseconds - start.uptimeNanoseconds
            let timeInterval = Double(nanoTime) / 1_000_000
            
            strings += timeInterval
        }
        
        strings /= Double(repetitions)
        stringsPerformance = String(Int(strings)) + " ms"
    }
    
    func isPrime(n:Int) -> Bool
    {
        if n % 2 == 0 { return false }
        
        var i = 3
        while i*i<=n {
            i+=3
            if n % i == 0 { return false }
        }
        
        return true;
    }
    
    func isPandigital(n: Int) -> Bool
    {
        var index = n
        var count = 0
        var digits = 0
        var digit = 0
        var bit = 0
        
        repeat
        {
            digit = index % 10;
            if digit == 0 { return false }
            bit = 1 << digit
    
            let check = digits | bit
            if digits == check { return false }
            digits = check
    
            count += 1
            index /= 10
        } while (n > 0)
        
        return (1 << count) - 1 == digits >> 1;
    }
    
    func calculateFoo(n : Int) -> Double
    {
        var a = 1.0
        for _ in 0..<n
        {
            a = a * 1.123456789101112 / Double(n)
        }
        
        return a
    }
}

struct FooContainer
{
    public var _x = 0
    public var _y = 0
    
    public func Sum () -> Int
    {
        return _x + _y
    }
}
